### PR TITLE
[AutoDiff] Clean up.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Store `AutoDiffConfig` in `SILDifferentiabilityWitness` instead of storing
  the individual components. This makes it cheaper to get an `AutoDiffConfig`.
- Unify parsing logic and diagnostics.
- Minor style changes.

### CODE REVIEW

**Comment by dan-zheng** [comment]
Please hold off on review, more SIL parser changes are coming (code deduplication and necessary changes for https://github.com/apple/swift/pull/27719).

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
Merging to unblock progress.

**Pull request merged**
